app/buck2_execute/src/materialize/http.rs

```diff
@@ -99,8 +101,36 @@ impl AsHttpError for HttpDownloadError {
     }
 }
 
+struct MaterializerDispatchableHttpRetryWarning {}
+
+impl MaterializerDispatchableHttpRetryWarning {
+    fn new() -> Self {
+        Self {}
+    }
+}
+
+impl DispatchableHttpRetryWarning for MaterializerDispatchableHttpRetryWarning {
+    fn dispatch(&self, dur: &Duration, _retries: usize, url: &str) {
+        let event: HttpRequestRetried = HttpRequestRetried {
+            backoff_duration_secs: dur.as_secs(),
+            url: url.to_owned(),
+        };
+
+        match buck2_events::dispatch::get_dispatcher_opt() {
+            Some(dispatcher) => {
+                dispatcher.instant_event(event);
+            }
+            None => {
+                tracing::warn!("Failed to dispatch HttpRequestRetried event: {:?}", event)
```

Comment:
So here what we actually would like to call is this function:

```rust
pub fn broadcast_instant_event<E: Into<buck2_data::instant_event::Data> + Clone>(
    event: &E,
) -> bool {
    let mut has_subscribers = false;

    for cmd in ACTIVE_COMMANDS.lock().values() {
        cmd.dispatcher.instant_event(event.clone());
        has_subscribers = true;
    }

    has_subscribers
}
```

(as-is unfortunately the logging wouldn't work when using the deferred materializer)

This function lives in `buck2_server`, so what we need to do is allow this to be injected into the callsites. 

Maybe the easiest way to do this is to simply have the `HttpClient` provide its logging method (because this is already threaded through to all the places you need it), via a method like:

```
trait HttpClient {
  ..

  fn retry_logger(&self) -> &dyn DispatchableHttpRetryWarning;
}
```

You probably should actually keep the "log to stderr" implementation as just the default implementation for this method on `HttpClient`. This way all you need to do is just wrap the client being constructed for the server with the event-logging code.

So, putting it all together, what we need to do is:

- Move this code to `buck2_server`
- Have it implement the broadcast
- When we create the server's http client (in `DaemonState::init_data`), we wrap it with some new struct implementing HttpClient that uses the event-logging implementation
- When we create ManifoldClient's http client, you inject the other one in
- When calling http_retry you get the retry logging from `HttpClient`

I can also make some of those changes if you'd like, just let me know